File: db-query-backend/src/schema/schema.service.ts
import { Injectable } from '@nestjs/common';
import { ConnectionManagerService } from '../connections/connection-manager.service';

type TableSchema = {
  name: string;
  columns: string[];
};

@Injectable()
export class SchemaService {
  constructor(
    private readonly connectionManager: ConnectionManagerService,
  ) {}

  /* =========================
     GET TABLES / COLLECTIONS
     ========================= */
  async getTables(connectionId: string): Promise<string[]> {
    const conn =
      this.connectionManager.getConnection(connectionId);

    /* ---------- SQL ---------- */
    if (conn.kind === 'sql') {
      const ds = conn.ds;
      const dbType = ds.options.type;

      /* PostgreSQL */
      if (dbType === 'postgres') {
        const result = await ds.query(`
          SELECT table_name
          FROM information_schema.tables
          WHERE table_schema = 'public'
            AND table_type = 'BASE TABLE'
          ORDER BY table_name
        `);

        return result.map((r: any) => r.table_name);
      }

      /* MySQL */
      if (dbType === 'mysql') {
        const dbName = ds.options.database;

        const result = await ds.query(
          `
          SELECT table_name
          FROM information_schema.tables
          WHERE table_schema = ?
            AND table_type = 'BASE TABLE'
          ORDER BY table_name
        `,
          [dbName],
        );

        return result.map(
          (r: any) => r.TABLE_NAME,
        );
      }

      return [];
    }

    /* ---------- MongoDB ---------- */
    if (conn.kind === 'mongo') {
      const collections =
        await conn.db.listCollections().toArray();

      return collections.map((c) => c.name);
    }

    return [];
  }

  /* =========================
     GET COLUMNS / FIELDS
     ========================= */
  async getColumns(
    connectionId: string,
    table: string,
  ) {
    const conn =
      this.connectionManager.getConnection(connectionId);

    /* ---------- SQL ---------- */
    if (conn.kind === 'sql') {
      const ds = conn.ds;
      const dbType = ds.options.type;

      /* PostgreSQL */
      if (dbType === 'postgres') {
        return ds.query(
          `
          SELECT column_name, data_type
          FROM information_schema.columns
          WHERE table_name = $1
            AND table_schema = 'public'
          ORDER BY ordinal_position
        `,
          [table],
        );
      }

      /* MySQL */
      if (dbType === 'mysql') {
        const dbName = ds.options.database;

        const result = await ds.query(
          `
          SELECT COLUMN_NAME, DATA_TYPE
          FROM information_schema.columns
          WHERE table_schema = ?
            AND table_name = ?
          ORDER BY ORDINAL_POSITION
        `,
          [dbName, table],
        );

        return result.map((r: any) => ({
          column_name: r.COLUMN_NAME,
          data_type: r.DATA_TYPE,
        }));
      }

      return [];
    }

    /* ---------- MongoDB ---------- */
    if (conn.kind === 'mongo') {
      const collection = conn.db.collection(table);

      const docs = await collection
        .find({})
        .limit(10)
        .toArray();

      const fieldSet = new Set<string>();

      docs.forEach((doc) => {
        Object.keys(doc).forEach((key) =>
          fieldSet.add(key),
        );
      });

      return Array.from(fieldSet).map((field) => ({
        column_name: field,
        data_type: 'mixed',
      }));
    }

    return [];
  }

  /* =========================
     FULL SCHEMA (SQL ONLY)
     ========================= */
  async getFullSchema(
    connectionId: string,
  ): Promise<TableSchema[]> {
    const conn =
      this.connectionManager.getConnection(connectionId);

    // NL schema is SQL-only
    if (conn.kind !== 'sql') {
      return [];
    }

    const tables = await this.getTables(connectionId);

    const schema: TableSchema[] = [];

    for (const table of tables) {
      const columns =
        await this.getColumns(connectionId, table);

      schema.push({
        name: table,
        columns: columns.map(
          (c: any) => c.column_name,
        ),
      });
    }

    return schema;
  }
}

File: db-query-backend/app/services/schema_service.py
"""Port of schema/schema.service.ts"""
from sqlalchemy import text

from app.models.dto import TableSchema
from app.services.connection_manager import connection_manager, SqlConnection, MongoConnection


class SchemaService:
    def get_tables(self, connection_id: str) -> list[str]:
        conn = connection_manager.get(connection_id)

        if isinstance(conn, MongoConnection):
            return conn.db.list_collection_names()

        # SQL
        if conn.db_type == "postgres":
            query = text(
                "SELECT table_name FROM information_schema.tables "
                "WHERE table_schema = 'public' "
                "AND table_type = 'BASE TABLE' "
                "ORDER BY table_name"
            )
        elif conn.db_type == "mysql":
            query = text(
                "SELECT table_name FROM information_schema.tables "
                "WHERE table_schema = DATABASE() "
                "AND table_type = 'BASE TABLE' "
                "ORDER BY table_name"
            )
        else:
            raise ValueError(f"getTables not supported for {conn.db_type}")

        with conn.engine.connect() as c:
            return [row[0] for row in c.execute(query)]

    def get_columns(self, connection_id: str, table: str) -> list[dict]:
        conn = connection_manager.get(connection_id)

        if isinstance(conn, MongoConnection):
            # Same "infer from a sample" approach as the original
            docs = list(conn.db[table].find().limit(10))
            fields: dict[str, set[str]] = {}
            for doc in docs:
                for key, value in doc.items():
                    fields.setdefault(key, set()).add(type(value).__name__)
            return [{"name": k, "type": "/".join(sorted(v))} for k, v in fields.items()]

        if conn.db_type == "postgres":
            query = text(
                "SELECT column_name, data_type FROM information_schema.columns "
                "WHERE table_schema = 'public' AND table_name = :table "
                "ORDER BY ordinal_position"
            )
        elif conn.db_type == "mysql":
            query = text(
                "SELECT column_name, data_type FROM information_schema.columns "
                "WHERE table_schema = DATABASE() AND table_name = :table "
                "ORDER BY ordinal_position"
            )
        else:
            raise ValueError(f"getColumns not supported for {conn.db_type}")

        with conn.engine.connect() as c:
            return [
                {"column_name": row[0], "data_type": row[1]}
                for row in c.execute(query, {"table": table})
            ]

    def get_full_schema(self, connection_id: str) -> list[TableSchema]:
        """SQL-only, same as the original."""
        conn = connection_manager.get(connection_id)
        if isinstance(conn, MongoConnection):
            raise ValueError("getFullSchema is SQL-only")

        tables = self.get_tables(connection_id)
        result = []
        for table in tables:
            cols = self.get_columns(connection_id, table)
            result.append(TableSchema(name=table, columns=[c["column_name"] for c in cols]))
        return result


schema_service = SchemaService()
